File: js/show.js
function readCvs() {
	Plotly.d3.csv("arq.csv", function(data){
		processData(data)
	});
}

function processData(allRows) {

	var csvdata = {}	
	var arqwin = ['srx','stx','etx','erx']

	for (var i=0; i<allRows.length; i++) {
		row = allRows[i];
		ts = row['ts'];
		node = row['node'];
		if (!(ts in csvdata)) {
		   csvdata[ts] = {};
		}

		for (var j = 0; j < arqwin.length; j++) {
			var key = arqwin[j];		
			if (!(key in csvdata[ts])) {
			   csvdata[ts][key] = {};
			}
			csvdata[ts][key][node] = row[key];
		}
	}

	createTraces(csvdata)

}

window.traces = {};

function createTraces(csvdata) {
	
	for (var ts in csvdata) {
		var fr = 'frame' + ts.toString();
		var xs = [];
		var ys = [];
		for (var ak in csvdata[ts])
		{
			var y = [];
			var x = [];
			for (var node in csvdata[ts][ak])		
			{
				x.push(csvdata[ts][ak][node]);
				y.push(node);
			}
			xs.push(x);
			ys.push(y);
		}
		
		traces[ts] = [];
		for (var j=0; j < xs.length; j++)
		{
			traces[ts].push({x: xs[j], y: ys[j], mode: 'markers', type: 'scatter', marker: { size: 12 }});
		}
	}	


	createFrames(traces)

}

window.frames = [];

function createFrames(traces)
{
	for(var ts in traces)
	{
		var frame = 'frame' + ts.toString();
		frames.push({data: traces[ts], name: frame});
	}
	createSteps(traces)
}

window.allSteps = [];

function createSteps(traces)
{
	for(var ts in traces)
	{
		var frame = 'frame' + ts.toString();
		allSteps.push({
		  label: ts.toString(),
		  method: 'animate',
		  args: [[frame], {
		      mode: 'immediate',
		      frame: {redraw: false, duration: 500},
		      transition: {duration: 500}
		    }]
		});
	}

}


function doPlot() {

	Plotly.plot('myDiv', Object.values(traces)[0],
 	{
	  updatemenus: [{
		direction: 'left',
        pad: {r: 10, t: 87},
        showactive: false,
        type: 'buttons',
        x: 0.1,
        xanchor: 'right',
        y: 0,
        yanchor: 'top',
        buttons: [
            {
                args: [null, {frame: {duration: 500, redraw: false},
									fromcurrent: true, 
									transition: {duration: 300, easing: 'quadratic-in-out'}
								}
						],
                label: 'Play',
                method: 'animate'
            },
			{
                args: [null, {frame: {duration: 100, redraw: false},
									fromcurrent: true, 
									transition: {duration: 50, easing: 'quadratic-in-out'}
								}
						],
                label: 'Play fast',
                method: 'animate'
            },
            {
                args: [[null], {frame: {duration: 0, redraw: false}, 
									mode: 'immediate',
						            transition: {duration: 0}
								}
						],
                label: 'Pause',
                method: 'animate'
            }
        ]        
    }]
	},
    {
	  xaxis: {autorange: true},
	  yaxis: {autorange: true}
	}).then(function () {
	  Plotly.addFrames('myDiv', frames);
	})

}

window.frame_c = 0;

function clickNext() {
 /*
	gd = helpers.getGraphDiv('myDiv');

	var currentFrame = gd._fullLayout._currentFrame;
	var nextFrame;
	if(currentFrame) {
		var idx = -1;
		for(i = 0; i < frames.length; i++) {
		    frame = frames[i];
		    if(frame.type === 'byname' && frame.name === currentFrame) {
		        idx = i;
		        break;
		    }
		}

		if(idx > 0 && idx < frameList.length - 1) {
			nextFrame = frames[idx + 1];
		}
	}
	if(nextFrame)
	{
	 	Plotly.animate('myDiv', [nextFrame], {
		  xaxis: {autorange: true},
		  yaxis: {autorange: true},
		frame: [
		  {duration: 200},
		],
		transition: [
		  {duration: 200, easing: 'cubic-in-out'},
		],
		mode: 'afterall'
		})
	}
*/
/* frame_c++
  if(frame_c==frames.length)frame_c=0
  k = Object.keys(traces)[frame_c]
  var fr = 'frame' + k.toString()
 
  Plotly.animate('myDiv', [fr], {
	  xaxis: {autorange: true},
	  yaxis: {autorange: true},
    frame: [
      {duration: 200},
    ],
    transition: [
      {duration: 200, easing: 'cubic-in-out'},
    ],
    mode: 'afterall'
  })
*/
}

function clickPrev() {

}


///////////////////////////
readCvs();
setTimeout(function(){ 

console.log(frames)
console.log(allSteps.length)


console.log(Object.values(traces)[0])

doPlot();

    }, 3000);

File: python/show.py

import plotly

from plotly.offline import init_notebook_mode, iplot, plot
from IPython.display import display, HTML
import pandas as pd

init_notebook_mode(connected=True) # run at the start of every ipython notebook


url = 'https://raw.githubusercontent.com/plotly/datasets/master/gapminderDataFiveYear.csv'
dataset = pd.read_csv(url)

years = ['1952', '1962', '1967', '1972', '1977', '1982', '1987', '1992', '1997', '2002', '2007']
# make list of continents
continents = []
for continent in dataset['continent']:
    if continent not in continents:
        continents.append(continent)
# make figure
figure = {
    'data': [],
    'layout': {},
    'frames': [],
    'config': {'scrollzoom': True}
}

# fill in most of layout
figure['layout']['xaxis'] = {'range': [30, 85], 'title': 'Life Expectancy'}
figure['layout']['yaxis'] = {'title': 'GDP per Capita', 'type': 'log'}
figure['layout']['hovermode'] = 'closest'
figure['layout']['slider'] = {
    'args': [
        'slider.value', {
            'duration': 400,
            'ease': 'cubic-in-out'
        }
    ],
    'initialValue': '1952',
    'plotlycommand': 'animate',
    'values': years,
    'visible': True
}
figure['layout']['updatemenus'] = [
    {
        'buttons': [
            {
                'args': [None, {'frame': {'duration': 500, 'redraw': False},
                         'fromcurrent': True, 'transition': {'duration': 300, 'easing': 'quadratic-in-out'}}],
                'label': 'Play',
                'method': 'animate'
            },
            {
                'args': [[None], {'frame': {'duration': 0, 'redraw': False}, 'mode': 'immediate',
                'transition': {'duration': 0}}],
                'label': 'Pause',
                'method': 'animate'
            }
        ],
        'direction': 'left',
        'pad': {'r': 10, 't': 87},
        'showactive': False,
        'type': 'buttons',
        'x': 0.1,
        'xanchor': 'right',
        'y': 0,
        'yanchor': 'top'
    }
]

figure['layout']['updatemenus'][0]['buttons'].append(dict(label='>',method='animate',args=[['1952','1962'], {'frame': {'duration': 500, 'redraw': False}, 'mode': 'fromcurrent','transition': {'duration': 300, 'easing': 'quadratic-in-out'}}]))

sliders_dict = {
    'active': 0,
    'yanchor': 'top',
    'xanchor': 'left',
    'currentvalue': {
        'font': {'size': 20},
        'prefix': 'Year:',
        'visible': True,
        'xanchor': 'right'
    },
    'transition': {'duration': 300, 'easing': 'cubic-in-out'},
    'pad': {'b': 10, 't': 50},
    'len': 0.9,
    'x': 0.1,
    'y': 0,
    'steps': []
}

# make data
year = 1952
for continent in continents:
    dataset_by_year = dataset[dataset['year'] == year]
    dataset_by_year_and_cont = dataset_by_year[dataset_by_year['continent'] == continent]

    data_dict = {
        'x': list(dataset_by_year_and_cont['lifeExp']),
        'y': list(dataset_by_year_and_cont['gdpPercap']),
        'mode': 'markers',
        'text': list(dataset_by_year_and_cont['country']),
        'marker': {
            'sizemode': 'area',
            'sizeref': 200000,
            'size': list(dataset_by_year_and_cont['pop'])
        },
        'name': continent
    }
    figure['data'].append(data_dict)
    
# make frames
for year in years:
    frame = {'data': [], 'name': str(year)}
    for continent in continents:
        dataset_by_year = dataset[dataset['year'] == int(year)]
        dataset_by_year_and_cont = dataset_by_year[dataset_by_year['continent'] == continent]

        data_dict = {
            'x': list(dataset_by_year_and_cont['lifeExp']),
            'y': list(dataset_by_year_and_cont['gdpPercap']),
            'mode': 'markers',
            'text': list(dataset_by_year_and_cont['country']),
            'marker': {
                'sizemode': 'area',
                'sizeref': 200000,
                'size': list(dataset_by_year_and_cont['pop'])
            },
            'name': continent
        }
        frame['data'].append(data_dict)

    figure['frames'].append(frame)
    slider_step = {'args': [
        [year],
        {'frame': {'duration': 300, 'redraw': False},
         'mode': 'immediate',
       'transition': {'duration': 300}}
     ],
     'label': year,
     'method': 'animate'}
    sliders_dict['steps'].append(slider_step)


print(type(figure['layout']['updatemenus'][0]) is dict )




    
figure['layout']['sliders'] = [sliders_dict]

plot(figure)

